src/webview/js/js.js

```diff
@@ -295,12 +295,23 @@ function lastMessage(): ?Element {
   return walkToMessage(documentBody.lastElementChild, 'previousElementSibling');
 }
 
-/** The minimum height (in px) to see of a message to call it visible. */
-const minOverlap = 20;
-
 function isVisible(element: Element, top: number, bottom: number): boolean {
   const rect = element.getBoundingClientRect();
-  return top + minOverlap < rect.bottom && rect.top + minOverlap < bottom;
+  return top < rect.bottom && rect.top < bottom;
+}
+
+// The number of pixels of the message that are allowed to be scrolled off the
+// bottom of the screen when we mark the message as "read". Picked
+// based on the bottom padding of messages in CSS. This doesn't work so great
+// in cases where messages have the "edited" badge or reaction emojis, but it's
+// more important for the single-line message case to not get inadvertently read
+// than it is for messages to get marked read exactly when the bottom of the
+// text of the message comes onto the screen. In the future, we may want more
+// complicated behaviour here, but this should be fine for now.
+const messageReadSlop = 16;
+
+function isRead(element: Element, top: number, bottom: number): boolean {
+  return bottom + messageReadSlop >= element.getBoundingClientRect().bottom;
```

Comment:
IIUC this function's jsdoc might look like:
```js
/**
 * A visible message is read when its bottom isn't too far down out of view.
 *
 * See `messageReadSlop` for specifically how far the bottom might be. The
 * idea is that a message becomes read when the bottom of its content
 * scrolls into view, excluding the padding between messages.
 *
 * This function doesn't check that the message is visible; it probably
 * makes sense to call only when `isVisible` is already known to be true.
 *
 * See `isVisible` for the meaning of `top` and `bottom`.
 */
function isRead(element: Element, top: number, bottom: number): boolean {
```

Does that agree with your understanding?

One thing this brings to mind is that, suppose we do find ourselves scrolled down into the middle of the unread history; then I believe with these changes, as you scroll up, we'll mark a message as read as soon as the first px of its bottom (in fact of its bottom padding) scrolls into view. With the existing code, we'll only do so once 20px scroll into view (including the padding, I guess -- so once the whole bottom padding plus 4px of the message's inner box are in view.)

Probably both those behaviors are almost equally annoying, plus that's a situation that I think we should avoid getting into (but there are almost certainly bugs that do get us into it), so this isn't a reason to block merging this if we don't see an easy way to avoid it.

Reply:
Yep, that's correct. Added a jsdoc.